qubes-rpc/qubes.DesktopNotify

```diff
@@ -0,0 +1,320 @@
+        signature="susssasa{sv}i"
+    )
+
+def notify(vm, trusted_notification):
+    notification = trusted_notification
+
+    #the ID is the caller ID --> fix it
+    our_id = IDS.get(notification.notification_id, 0)
+
+    #give the sending VM a dedicated field (summary) to hopefully avoid spoofing attacks (e.g. with backspaces)
+    summary = vm
+    if notification.body and notification.summary:
+        body = " : ".join([notification.summary, notification.body])
+    elif notification.body:
+        body = notification.body
+    else:
+        body = notification.summary
+
+    notification = Notification(
+        app_name=notification.app_name,
+        notification_id=our_id,
+        app_icon=notification.app_icon,
+        summary=summary,
+        body=body,
+        actions=notification.actions,
+        hints=notification.hints,
+        expire_timeout=notification.expire_timeout,
+    )
+
+    LOG.info(f"Notification from VM {vm}: {notification}")
+    ret_id = _notify(notification)
+
+    if trusted_notification.notification_id > 0:
+        IDS[trusted_notification.notification_id] = ret_id
+
+def close_notification(trusted_id):
+    #the ID is the caller ID --> fix it
+    try:
+        our_id = IDS.pop(trusted_id)
+    except KeyError:
+        LOG.warning("Received a closed request for a notification that's not in our databse. This is unexpected. Ignoring...")
+        return
+
+    iface = get_iface()
+    iface.CloseNotification(our_id, signature="u")
+
+def regex_sanitize(vm, regex, untrusted_str, field):
+    m = re.match(regex, untrusted_str)
+    if m:
+        return m.group(1)
+    raise ParseException(f"Received unexpected message content in the field {field} from VM {vm}.")
+
+def parse_notification_id(vm, untrusted):
+    return int(regex_sanitize(vm, r"^([0-9]{1,6})$", untrusted, "notification_id"))
+
+def parse_expire_timeout(vm, untrusted):
+    return int(regex_sanitize(vm, r"^(\-?[0-9]{1,10})$", untrusted, "expire_timeout"))
+
+def assert_printable(vm, untrusted, max_len):
+    l = len(untrusted)
+    if l >= max_len:
+        raise ParseException(f"Received content of excessive length ({l} >= {max_len}) from VM {vm}.")
+
+    allowed_chars = set(["\t", "\n", "\r"])
+    ret = []
+    for ch in untrusted:
+        cat = unicodedata.category(ch)
+        if cat[0] == "C" and ch not in allowed_chars:
+            #https://www.unicode.org/reports/tr44/#GC_Values_Table
+            ret += "?"
+        else:
+            ret += ch
+
+    ret = "".join(ret)
+
+    if not ALLOW_UNICODE:
+        ret = ret.encode(encoding="ascii", errors="replace").decode("ascii")
+
+    return ret
+
+def parse_app_name(vm, untrusted):
+    return assert_printable(vm, untrusted, 50)
+
+def parse_app_icon(vm, untrusted):
+    # The sending VM is supposed to either send a file:// URI or a well-known icon name.
+    # We don't support raw binary icons at the moment (and it might not be wise to do so).
```

Comment:
Transferring byte arrays is not an issue. Displaying however likely involves some buggy image library.

And as long as the VM template is the same, it is likely to have the same file in the receiving VM. Otherwise it's ignored, yes.

Reply:
> Transferring byte arrays is not an issue. Displaying however likely involves some buggy image library.

Qubes OS actually assumes the contrary.  Specifically, app icons are displayed in the GUI qube, and can be arbitrary RGBA pictures (bounded only in size).  If you were referring to PNG, JPEG, or other non-trivial image formats, I would agree with you, but RGBA is incredibly simple and hard to screw up.

Reply:
> Transferring byte arrays is not an issue. Displaying however likely involves some buggy image library.
> 
> And as long as the VM template is the same, it is likely to have the same file in the receiving VM. Otherwise it's ignored, yes.

The template is by no means guaranteed to be the same, and I suspect it often will not be.